docling_core/types/doc/document.py

```diff
@@ -1146,133 +1150,142 @@ def export_to_markdown(  # noqa: C901
-
-                    # Rendered the item
-                    if not strict_text:
-                        md_table = item.export_to_markdown()
-                        if md_table:
-                            parts.append(item.export_to_markdown())
-
-                    # Combine parts
-                    markdown_text = "\n".join(parts)
-
-                elif isinstance(item, PictureItem) and item_type in labels:
-                    parts = []
+            # Handle newlines between different types of content
+            if (
+                len(mdtexts) > 0
+                and not isinstance(item, (ListItem, GroupItem))
+                and in_list
+            ):
+                mdtexts[-1] += "\n"
+                in_list = False
+
+            if isinstance(item, GroupItem) and item.label in [
+                GroupLabel.LIST,
+                GroupLabel.ORDERED_LIST,
+            ]:
+                # Increment list nesting level when entering a new list
+                list_nesting_level += 1
+                in_list = True
+                continue
 
-                    # Compute the caption
-                    if caption := item.caption_text(self):
-                        parts.append(caption)
-                        parts.append("\n")
+            elif isinstance(item, GroupItem):
+                continue
 
-                    # Rendered the item
-                    if not strict_text:
-                        parts.append(f"{image_placeholder}")
+            elif isinstance(item, TextItem) and item.label in [DocItemLabel.TITLE]:
+                in_list = False
+                marker = "" if strict_text else "#"
+                text = f"{marker} {item.text}\n"
+                mdtexts.append(text.strip())
+
+            elif (
+                isinstance(item, TextItem)
+                and item.label in [DocItemLabel.SECTION_HEADER]
+            ) or isinstance(item, SectionHeaderItem):
+                in_list = False
+                marker = ""
+                if not strict_text:
+                    marker = "#" * level
+                    if len(marker) < 2:
+                        marker = "##"
+                text = f"{marker} {item.text}\n"
+                mdtexts.append(text.strip() + "\n")
+
+            elif isinstance(item, TextItem) and item.label in [DocItemLabel.CODE]:
+                in_list = False
+                text = f"```\n{item.text}\n```\n"
+                mdtexts.append(text)
+
+            elif isinstance(item, TextItem) and item.label in [DocItemLabel.CAPTION]:
+                # captions are printed in picture and table ... skipping for now
+                continue
 
-                    # Combine parts
-                    markdown_text = "\n".join(parts)
+            elif isinstance(item, ListItem) and item.label in [DocItemLabel.LIST_ITEM]:
+                in_list = True
+                # Calculate indent based on list_nesting_level
+                indent = "    " * (
```

Comment:
Instead of using literals like `"    "` (hard to read), I would rather include an indent argument like `json.dump` for the base indent and use that, e.g.:

```
def export_to_markdown(
  # ...
  base_indent: int = 4,
):
  # ...
  indent = (list_nesting_level - 1) * base_indent * " "
  # ...
```

Reply:
I tend to agree

Reply:
More so, any indentation of less than 4 is not legal and may render broken on some Markdown visualizers.